src/encoding.rs

```diff
@@ -2,67 +2,67 @@
 //! All methods err using VMError::FormatError for convenience.
 
 use byteorder::{ByteOrder, LittleEndian};
+use core::ops::Range;
 use curve25519_dalek::ristretto::CompressedRistretto;
 use curve25519_dalek::scalar::Scalar;
 
 use crate::errors::VMError;
 
-/// Reads a byte and returns the subsequent slice
-pub fn read_u8<'a>(slice: &'a [u8]) -> Result<(u8, &'a [u8]), VMError> {
-    if slice.len() < 1 {
-        return Err(VMError::FormatError);
+#[derive(Copy,Clone,Debug)]
+pub struct SliceView<'a> {
+    whole: &'a [u8],
+    index: usize,
+}
+
+impl<'a> SliceView<'a> {
+    // next advances the internal range and returns the 
+    // updated SliceView
+    fn next(self, offset: usize) -> Result<Self, VMError> {
+        self.index = self.index+offset;
+        Ok(self)
+    }
+
+    fn slice(self, offset: usize) -> Result<&'a [u8], VMError> {
+        self.whole.get(self.index..self.index+offset).ok_or(VMError::FormatError)
     }
-    Ok((slice[0], &slice[1..]))
 }
 
-/// Reads the LE32-encoded integer and returns the subsequent slice
-pub fn read_u32<'a>(slice: &'a [u8]) -> Result<(u32, &'a [u8]), VMError> {
-    if slice.len() < 4 {
-        return Err(VMError::FormatError);
+pub fn read_next_u8<'a>(slice: SliceView<'a>) -> Result<(u8, SliceView<'a>), VMError> {
+        let bytes = slice.slice(1)?;
+        Ok((bytes[0], slice.next(1)?))
     }
-    let x = LittleEndian::read_u32(slice);
-    Ok((x, &slice[4..]))
+
+pub fn read_next_u32<'a>(slice: SliceView<'a>) -> Result<(u32, SliceView<'a>), VMError> {
+    let bytes = slice.slice(4)?;
+    let x = LittleEndian::read_u32(bytes);
+    Ok((x, slice.next(4)?))
 }
 
-/// Reads the LE32-encoded integer as `usize` and returns the subsequent slice
-pub fn read_usize<'a>(slice: &'a [u8]) -> Result<(usize, &'a [u8]), VMError> {
-    let (n, rest) = read_u32(slice)?;
-    Ok((n as usize, rest))
+pub fn read_next_usize<'a>(slice: SliceView<'a>) -> Result<(usize, SliceView<'a>), VMError> {
+    let (n, next_slice) = read_next_u32(slice)?;
+    Ok((n as usize, next_slice))
 }
 
-/// Reads a 32-byte array and returns the subsequent slice
-pub fn read_u8x32<'a>(slice: &'a [u8]) -> Result<([u8; 32], &'a [u8]), VMError> {
-    if slice.len() < 32 {
-        return Err(VMError::FormatError);
-    }
+pub fn read_next_u8x32<'a>(slice: SliceView<'a>) -> Result<([u8; 32], SliceView<'a>), VMError> {
     let mut buf = [0u8; 32];
-    let (a, rest) = slice.split_at(32);
-    buf[..].copy_from_slice(a);
-    Ok((buf, rest))
+    let bytes = slice.slice(32)?;
```

Comment:
We use a couple of times a combination of `.slice($N) ... .next($N)` which is less safe than a method like `.split_at($N)` that simultaneously gives a $N-byte slice and the remainder of the SliceView. Or maybe two sliceviews, where the first one is convertible to normal slice via Deref trait.

Reply:
However, the code would be simpler if the "remainder" was the in-place update, so we wouldn't have to re-shadow the same variable and worry that we accidentally use a wrong copy of a slice, messing up parsing.

Something like `.read(&mut self, n: usize) -> Result<SliceView, VMError>` which would slice off `n` bytes and advance the internal offset.

Then we'd need to have `end: usize` field in addition to `offset: usize` to know where the sliceview ends.